mindone/models/modules/flash_attention.py

```diff
@@ -0,0 +1,114 @@
+        input_layout (str): The input data layout. Defaults to "BNSD".
+        high_precision (bool): Determines whether to use high precision mode for attention calculations. Defaults to True.
+    Attributes:
+        use_new_flash_attention (bool): Indicates whether the new FlashAttention module supported in ms 2.3.0.
+        flash_attention (FlashAttention): An instance of the FlashAttention module used for attention calculations.
+        fa_mask_dtype (dtype): The data type used for the attention mask (ms.uint8 or ms.float16 depending on the version).
+        fix_head_dims (list): A list of integers representing head dimensions to be padded to 2**n * 64.
+    """
+
+    def __init__(
+        self,
+        head_dim: int,
+        head_num: int,
+        fix_head_dims: Optional[List[int]] = None,
+        attention_dropout: float = 0.0,
+        input_layout: str = "BNSD",
+        high_precision: bool = True,
+    ):
+        super().__init__()
+        assert FLASH_IS_AVAILABLE, "FlashAttention is not Available!"
+        self.use_new_flash_attention = MSVersion >= "2.3.0"
+        if self.use_new_flash_attention:
+            self.flash_attention = FlashAttention(
+                scale_value=1.0 / math.sqrt(head_dim),
+                head_num=head_num,
+                input_layout=input_layout,
+                keep_prob=1 - attention_dropout,
+            )
+        else:
+            self.flash_attention = FlashAttention(
+                head_dim=head_dim,
+                head_num=head_num,
+                high_precision=high_precision,
+                dropout_rate=attention_dropout,
+            )  # TODO: how high_precision affect the training or inference quality
+        self.fa_mask_dtype = choose_flash_attention_dtype()  # ms.uint8 or ms.float16 depending on version
+        self.fix_head_dims = fix_head_dims  # A list of integers to be mapped to 2**n * 64
+        if self.fix_head_dims is not None and not isinstance(self.fix_head_dims, (list, tuple)):
+            self.fix_head_dims = [self.fix_head_dims]
+
+    def construct(self, q, k, v, mask=None):
+        q_b, h, q_n, d = q.shape  # (b, h, n, d)
+        head_dim = d
+        if mask is None:
+            mask = ops.zeros((q_b, q_n, q_n), self.fa_mask_dtype)
+        #   a trick to pad head dimensions to 2**n * 64
+        if self.fix_head_dims is not None and head_dim in self.fix_head_dims:
+            # pad to 2**n * 64 to avoid accuracy errors
+            padding_size = 64 * 2 ** math.ceil(math.log(head_dim / 64, 2)) - head_dim
+            q = msnp.pad(q, ((0, 0), (0, 0), (0, 0), (0, padding_size)), constant_value=0)
+            k = msnp.pad(k, ((0, 0), (0, 0), (0, 0), (0, padding_size)), constant_value=0)
+            v = msnp.pad(v, ((0, 0), (0, 0), (0, 0), (0, padding_size)), constant_value=0)
+        if self.use_new_flash_attention:
+            out = self.flash_attention(
+                q.to(ms.float16),
```

Comment:
I added an argument and an attribute named `dtype`, which could allow setting data type of q, k, and v.

Actually, I only know FA works with ms.float16. I don't know if it can work with `ms.float32` or `ms.bfloat16`.